modules/message-lane/src/lib.rs

```diff
@@ -238,6 +243,9 @@ decl_module! {
 				Error::<T, I>::InvalidMessagesProof
 			})?;
 
+			// try to decode message payloads
+			let messages: Vec<_> = messages.into_iter().map(Into::into).collect();
```

Comment:
With #384 implemented, we should move the decoding part after weight lower bound checking and add an extra step to compare the declared weight with the actual one.